### PR TITLE
Refactor how connection restarting is handled

### LINKED ISSUE #14327: Code Health: follow ups from connection restarting
From #14060. See also https://github.com/microsoft/terminal/pull/14060#pullrequestreview-1165989506

The control shouldn't be responsible for restarting the connection itself. It should request a new one from the app. 

* [ ] Move [this code](https://github.com/sotteson1/terminal/blob/62888a5a0f2ccda00b53e52e96537fd67213e4ab/src/cascadia/TerminalControl/ControlCore.cpp#L415-L420) to the page, and bubble an event to request a new connection
* [ ] Revert [this](https://github.com/sotteson1/terminal/blob/62888a5a0f2ccda00b53e52e96537fd67213e4ab/src/cascadia/TerminalConnection/ConnectionStateHolder.h#L89-L95)
* [ ] Investigate only clearing the buffer to scrollback, instead of `inheritCursor`ing

### PR DESCRIPTION
**Pull request by zadjii-msft**
Admittedly, this is a lot of annoying plumbing. 

We didn't love that a connection could transition back in the state diagram, from Closed -> Start. That felt wrong. To remedy this, we're going to allow the ControlCore to re-create connections, all on its own. This builds on a lot of work that was originally done for Process Model v2 (#5000), and was originally slated to be merged in #12938. 

Basically, all the properties for a Connection are stashed already in a `ConnectionState` object. We're gonna take that, and stash the info in the ControlCore. When the control wants to restart the connection, it just uses that to rebuild a new one. 

Cleanup from #14060 

Closes #14327

### CODE REVIEW

File: src/cascadia/TerminalControl/TermControl.idl [hunk 1]
```diff
@@ -29,6 +29,8 @@ namespace Microsoft.Terminal.Control
         void UpdateControlSettings(IControlSettings settings);
         void UpdateControlSettings(IControlSettings settings, IControlAppearance unfocusedAppearance);
 
+        Microsoft.Terminal.TerminalConnection.ConnectionInformation ConnectionInfo{ get; set; };
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.idl:32]: format this!

Comment (DHowett) [reply src/cascadia/TerminalControl/TermControl.idl:32]: (manually; the `{` needs a space before it)

File: src/cascadia/TerminalControl/TermControl.h [hunk 2]
```diff
@@ -33,6 +33,9 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         winrt::fire_and_forget UpdateControlSettings(Control::IControlSettings settings, Control::IControlAppearance unfocusedAppearance);
         IControlSettings Settings() const;
 
+        TerminalConnection::ConnectionInformation ConnectionInfo();
+        void ConnectionInfo(TerminalConnection::ConnectionInformation info);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.h:37]: nit: these guys should be `const&`s

File: src/cascadia/TerminalControl/ControlCore.h [hunk 3]
```diff
@@ -326,6 +327,8 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         bool _isBackgroundTransparent();
         void _focusChanged(bool focused);
 
+        void _setConnection(TerminalConnection::ITerminalConnection connection);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.h:330]: nit: should be a `const&` unless you `std::move` from it

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 4]
```diff
@@ -1300,35 +1299,25 @@ namespace winrt::TerminalApp::implementation
...
-        TraceLoggingWrite(
-            g_hTerminalAppProvider,
-            "ConnectionCreated",
-            TraceLoggingDescription("Event emitted upon the creation of a connection"),
-            TraceLoggingGuid(connectionType, "ConnectionTypeGuid", "The type of the connection"),
-            TraceLoggingGuid(profile.Guid(), "ProfileGuid", "The profile's GUID"),
-            TraceLoggingGuid(sessionGuid, "SessionGuid", "The WT_SESSION's GUID"),
-            TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
-            TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage));
+        return TerminalConnection::ConnectionInformation(className, connectionSettings);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1315]: nit: use `{}` initialization syntax, otherwise this looks like a function call

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 5]
```diff
@@ -2704,10 +2693,44 @@ namespace winrt::TerminalApp::implementation
...
         {
             connection.Resize(controlSettings.DefaultSettings().InitialRows(), controlSettings.DefaultSettings().InitialCols());
+
+            // If we had an existing connection (i.e., defterm), then get the
+            // current settings out of it and stash those with the control. It
+            // won't be perfect, but it'll be good enough.
+            if (auto conpty{ existingConnection.try_as<ConptyConnection>() })
+            {
+                connectionInfo = TerminalConnection::ConnectionInformation(winrt::name_of<TerminalConnection::ConptyConnection>(),
+                                                                           conpty.ToSettings());
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2715]: It feels like ConptyConnection (or, any connection) should be able to produce a fully-formed `ConnectionInformation` itself, without the consumer needing to know about its classname

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 6]
```diff
@@ -2721,10 +2744,15 @@ namespace winrt::TerminalApp::implementation
             if (bothAltsPressed)
             {
                 std::tie(connection, debugConnection) = OpenDebugTapConnection(connection);
+                // Debug Tap connections aren't really restartable. The
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2747]: so what happens if you _do_ restart one?

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 7]
```diff
@@ -257,6 +257,28 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
         return vs;
     }
 
+    // This is used for defterm connections. The app will get these from the
```
Comment (DHowett) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp:260]: NAK on a generic named method on a public interface being _for a specific use case only._ Either the name should be suitably nongeneric or the method's behavior suitable widened!

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 8]
```diff
@@ -276,6 +298,7 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
             {
                 _passthroughMode = winrt::unbox_value_or<bool>(settings.TryLookup(L"passthroughMode").try_as<Windows::Foundation::IPropertyValue>(), _passthroughMode);
             }
+            _inheritCursor = winrt::unbox_value_or<bool>(settings.TryLookup(L"inheritCursor").try_as<Windows::Foundation::IPropertyValue>(), _inheritCursor);
```
Comment (DHowett) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp:301]: clever

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 9]
```diff
@@ -331,6 +324,28 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
     }
 
+    // Method Description:
+    // - Setup our event handlers for this connection. If we've currently got a
+    //   connection, then this'll revoke the existing connection's handlers.
+    void ControlCore::_setConnection(TerminalConnection::ITerminalConnection connection)
+    {
+        if (_connection)
+        {
+            _connection.TerminalOutput(_connectionOutputEventToken);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:334]: should we go ahead and make this a revoker while we're here?

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 10]
```diff
@@ -392,11 +407,48 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+
+                    // Get our current size in rows/cols, and hook them up to
+                    // this connection too.
+                    {
+                        auto cx = gsl::narrow_cast<til::CoordType>(_panelWidth * _compositionScale);
+                        auto cy = gsl::narrow_cast<til::CoordType>(_panelHeight * _compositionScale);
+                        cx = std::max(cx, _actualFont.GetSize().width);
+                        cy = std::max(cy, _actualFont.GetSize().height);
+                        const auto viewInPixels = Viewport::FromDimensions({ 0, 0 }, { cx, cy });
+                        const auto vp = _renderEngine->GetViewportInCharacters(viewInPixels);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:434]: yikes, we don't know our size any other way???

**Review by DHowett — COMMENTED** [review]
Okay, some big questions. I don't love how we're becoming _more stringly-typed_ but perhaps it offers us some heretofore unobserved opportunities...

File: src/cascadia/TerminalControl/ControlCore.cpp (hunk 10 above)
Comment (lhecker) [reply src/cascadia/TerminalControl/ControlCore.cpp:434]: Yes, this should use `_terminal->GetViewport()`.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 5 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2715]: That... might? be able to be done. Not sure it's immediately relevant here (in the defterm context). But I guess it could be `ConptyConnection::CreateConnectionInfo(settings...) -> ConnectionInfo`, and have it fill in its own name.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 6 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2747]: You can restart the underlying connection, but not the debugtap-ness of it. The debug spew just stops, but the real pane gets restarted.

File: src/cascadia/TerminalConnection/ConnectionInformation.cpp [hunk 11]
```diff
@@ -36,21 +43,58 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
+        ...\src\cascadia\TerminalConnection\ConnectionInformation.cpp(47)\TerminalConnection.dll!01234:
+        (caller: 01234) LogHr(1) tid(83a8) 800700A1 The specified
+        path is invalid.
+            [...TerminalConnection::implementation::ConnectionInformation::CreateConnection(RoActivateInstance(name,
+            raw))]
+        */
         //
-        // When we get to extensions (GH#4000), we may want to revisit.
-        if (LOG_IF_FAILED(RoActivateInstance(name, raw)))
+        // So to avoid those, we'll manually instantiate these
```
Comment (lhecker) [inline src/cascadia/TerminalConnection/ConnectionInformation.cpp:66]: This addition seems kind of unrelated. Is there an issue for this?

**Review by lhecker — APPROVED** [review]

File: src/cascadia/TerminalConnection/ConnectionInformation.cpp (hunk 11 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalConnection/ConnectionInformation.cpp:66]: I should just get rid of that comment honestly. I think that was due to some wacky out-of-proc stuff I was doing for PMv2, and since we're _not doing that_ anymore, I don't think we'll run into that anymore.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 12]
```diff
@@ -1300,35 +1299,25 @@ namespace winrt::TerminalApp::implementation
...
+                                                                                      newWorkingDirectory,
+                                                                                      settings.StartingTitle(),
+                                                                                      envMap.GetView(),
+                                                                                      settings.InitialRows(),
+                                                                                      settings.InitialCols(),
+                                                                                      winrt::guid());
 
-            sessionGuid = conhostConn.Guid();
-            connection = conhostConn;
+            connectionSettings.Insert(L"passthroughMode", Windows::Foundation::PropertyValue::CreateBoolean(settings.VtPassthrough()));
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1312]: Good catch

File: src/cascadia/TerminalControl/TermControl.h [hunk 13]
```diff
@@ -35,6 +35,9 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         winrt::fire_and_forget UpdateControlSettings(Control::IControlSettings settings, Control::IControlAppearance unfocusedAppearance);
         IControlSettings Settings() const;
 
+        TerminalConnection::ConnectionInformation ConnectionInfo();
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.h:38]: I don't totally understand why TerminalPage can't just _make a totally new connection out of nothing_ - I find it so weird that TermControl now has a Connection (which we gave it) and a ConnectionInfo (_which we used to make the Connection_, and then also gave it). How would a public user of TermControl's public API even know what to do with all of these members?

Comment (DHowett) [reply src/cascadia/TerminalControl/TermControl.h:38]: I don't love the pattern of storing effectively random app data on random other classes, you know? We did the same for Profile (and pushed it all the way down to the Core)...

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 14]
```diff
@@ -425,11 +439,43 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+                {
+                    // Manually update the startingDirectory. If we had a CWD set
+                    // by the client, then let's try and use that as the CWD for
+                    // a restart, so it's more "seamless"
+                    _ConnectionInfo.Settings().Insert(L"startingDirectory", Windows::Foundation::PropertyValue::CreateString(WorkingDirectory()));
+                    // pass in the magic "inheritCursor" setting to the
+                    // connection's settings. This'll cause conpty to restart
+                    // the connection at the current place in the buffer.
+                    _ConnectionInfo.Settings().Insert(L"inheritCursor", Windows::Foundation::PropertyValue::CreateBoolean(true));
+                    auto c = TerminalConnection::ConnectionInformation::CreateConnection(_ConnectionInfo);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:456]: okay, I see, but I still don't love it. "Sometimes the connection comes from the app and sometimes TermControl spontaneously ~~creates life and seeds it onto an empty planet~~ makes its own connection".

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 15]
```diff
@@ -425,11 +439,43 @@ namespace winrt::Microsoft::Terminal::Control::implementation
                 return true;
             }
 
-            if (ch == Enter)
+            // If we have a connection info with which to try and re-create the
+            // connection, then let's try doing that on Enter
+            if (_ConnectionInfo != nullptr)
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:444]: I guess this puts us in a really weird position, where `ConptyConnection` is responsible for printing out the "press ^D to restart" message, but TermControl is ultimately responsible for handling it.

But at the same time, this means that users of other connections (#4661) won't be told that ^D/Enter are options... but they MIGHT be? If they can be reconstructed via ConnectionInfo?

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 16]
```diff
@@ -364,6 +357,27 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
     }
 
+    // Method Description:
+    // - Setup our event handlers for this connection. If we've currently got a
+    //   connection, then this'll revoke the existing connection's handlers.
+    void ControlCore::_setConnection(const TerminalConnection::ITerminalConnection& connection)
+    {
+        if (_connection)
+        {
+            _connectionOutputEventRevoker.revoke();
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:367]: overwriting `_connectionStateChangedRevoker` later _should_ revoke it automatically, without you needing to check this

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 17]
```diff
@@ -263,6 +263,28 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
+    // connection. It likely won't be perfect - for example, any env vars that
+    // were used to create the initial client process will be lost in this
+    // process.
+    //
+    // Rows, Columns, guid - none of that should be persisted. That's irrelevant.
+    //
+    // Starting Directory - alas, we don't know that for a defterm connection.
+    // The Core might be able to fill it in, if the client app ever tells the
+    // Terminal.
+    Windows::Foundation::Collections::ValueSet ConptyConnection::GetDeftermSettings()
```
Comment (DHowett) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp:278]: for my own understanding
We have to use this in the magical scenario where we got a connection created out of nowhere and we had no way to inform TerminalPage what settings it started with?

**Review by DHowett — CHANGES_REQUESTED** [review]
14/15

**Comment by zadjii-msft** [comment]
Closed in favor of #15240
